### ISSUE TITLE
Developing module

### ISSUE THREAD

**Issue by 538Michael** [eXo-OpenSource/ml_pathfind#7]
Hello, I would like to edit the module, I tried to follow the instructions, but there's not .bat and .sln files, could you help please?

**Comment by jushar** [comment]
Open up a terminal and type `utils\premake5.exe vs2019`. This will generate the project files in the build directory.

**Comment by 538Michael** [comment]
Oh thank you, it worked. Another question, how could I build a .so file? I tried to Build the module, but only Win32 and x64 are available.

**Comment by jushar** [comment]
You need to clone the repository on a Linux machine (WSL would work too) and then execute:
```
utils/premake5 gmake
cd Build
make
```

**Comment by 538Michael** [comment]
It worked, thanks for your help and attention @Jusonex .

**Comment by 538Michael** [comment]
Oh, sorry again for bothering you, but I installed the module on my linux host and got this error:

> MODULE: Unable to load x64/modules/ml_pathfind.so (/lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by x64/modules/ml_pathfind.so))

**Comment by jushar** [comment]
That happens if you build the module on a system that is newer than the target system. The easiest way is to just build the module on the target system or alternatively setup a virtual machine/docker that uses the same OS image as your target host.

**Comment by 538Michael** [comment]
Worked. 👍

How did you convert `vehiclenodes.lua` from gps script to `sa_nodes.json`? 'Cause I've made some changes to `vehiclenodes.lua` and don't want to do again on `sa_nodes.json`. If possible ofc!

**Comment by jushar** [comment]
I honestly don't remember. I think I just used the editor's search&replace functionality. Writing a script that converts them shouldn't be too difficult either or you could use MTA's toJSON function.

**Comment by 538Michael** [comment]
@Jusonex any idea of how can I make an instantaneous path update while player is moving, but also lag-free? (Like GTA V)

**Comment by jushar** [comment]
You can do that with this module too as long as it's not updating the path too frequently and for too many players.
One path calculation only takes a few milliseconds. The network latency is usually higher.

For very frequent updates, you'd need a client-only solution. But that's currently not possible as MTA doesn't offer a way to  execute CPU-intensive tasks simultaneously. For that something like the implementation of https://github.com/multitheftauto/mtasa-blue/issues/918 is needed.

**Comment by 538Michael** [comment]
@Jusonex  At pathfind::Graph.cpp there's a comment called "TODO: Optimize this with index structures", what should be these index structures?

**Comment by jushar** [comment]
Currently, it iterates through all nodes to find the closest node (the starting node).

This is computationally pretty expensive as in the worst case it iterates as often as there are nodes in the list. This can be optimized with spatial data structures like [R trees](https://en.wikipedia.org/wiki/R-tree). These are for example also used in MTA for its colshape/marker calculations.

**Comment by jushar** [comment]
Actually, there might be a way to do it more efficiently on the client only.

You could use CEF to calculate the route and report the result back to Lua. This should be much faster as on the one hand JavaScript is JIT-compiled and on the other hand it runs in a secondary thread that doesn't block rendering.

**Comment by 538Michael** [comment]
I changed the A* to Dijkstra (I just need give it the destination node) and it returns to LUA the parents table. So I just need to call dijkstra once when the player marks the destination on the map and after it it's just update the path in the radar script using findNodeAt function to get current player node to get the path between current and destination. I tested and in my pc every dijkstra call is around 25ms and every findNodeAt call is around 2ms. So it's much better than calling A* every 50ms to update the path. Well, in my pc it's not lagging...

**So in a feel words:** Player request gps, Dijkstra is called, the module returns me **parentsTable**, **parentsTable** and **destinationNode** are sent to client side and a server side timer is started to call findNodeAt every 50ms to update player current position and send **currentNode** to client side, then client side radar script checks (#**parentsTable** > 0 and **currentNode** ~= nil), so we've a path to draw, then it's just draw it. 
(I made a sa_nodes.lua via sa_nodes.json to get the position of a node by id, to use it in draw function of the radar)

Tell me if you got my idea.
